Remove Tavern from API v3 list of guilds when 'guild' or 'publicGuilds' type parameter get added. (Fixes #12407)

AlexanderColen opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #12407

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
Removed the Tavern from showing up when fetching the list of guilds from API v3 and adding `type=guild` or `type=publicGuilds` as a query parameter.

Also added two new tests regarding this behaviour and fixed the failing tests as a result of this change.


[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 37028869-4c20-4ceb-8345-58b81b2619e8

AlexanderColen committed:
Excluding tavern from showing up in GET /groups API when 'guilds' or 'publicGuilds' type parameter is included.

AlexanderColen committed:
Fixed test errors.

AlexanderColen committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitica into api-tavern-exclusion

paglias commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -236,15 +236,27 @@ describe('GET /groups', () => {
     }).timeout(10000);
   });
 
+  it('makes sure that the tavern doesn\'t show up when guilds is passed as a query', async () => {
```
for this and the next test could we also add a check that `guilds.length` is `0` so that we're sure there's only `guilds[0]` to check?

AlexanderColen replied:
Because of previous tests, `guilds.length` at that moment is 4 and 2 for type `guilds` and `publicGuilds` respectively.

This seems to be because in the `before` some guilds already got created for other tests, like this one:
https://github.com/HabitRPG/habitica/blob/13a5b276e94e3e571e9a5d1fc641d5d0c82ac138/test/api/v3/integration/groups/GET-groups.test.js#L28-L40

So the two options are then to loop over every single guild and check if they don't contain `TAVERN_ID`, or to reset the database again using `await resetHabiticaDB();` to make sure that there are actually zero extra created guilds.

https://github.com/HabitRPG/habitica/blob/13a5b276e94e3e571e9a5d1fc641d5d0c82ac138/test/helpers/mongo.js#L36-L88

The first has the downside that it could make the tests take unnecessary long in the future, if more new guilds end up being created. The second might cause the tests afterwards to fail, so the tests would have to be moved to the end instead.

Which option would be the wisest to go for in this regard? Personally I think `await resetHabiticaDB();` is the cleaner option, since it makes sure that the Tavern documents becomes part of the guilds before the test.

paglias replied:
I'd go with the first option and loop over the resulting array to make sure the tavern isn't there, there are few guilds and even if some are added it would take thousands of results to slow down the tests.

```
expect(guilds.find(g => g.id === TAVERN_ID)).to.be.undefined
```

^ something like this should work!

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -344,7 +345,14 @@ api.getGroups = {
 
     if (req.query.member) {
-      filters._id = { $in: user.guilds };
+      // Includes 'guilds' nor 'publicGuilds' type.
```
can we move the check to `Group.getGroups ` here https://github.com/HabitRPG/habitica/blob/develop/website/server/models/group.js#L279 so that the tavern is never returned when that method is used instead of just in this route?

AlexanderColen replied:
_Move_ the check or keep it here and _add_ the check to `Group.getGroups` as well?

paglias replied:
I'd move the check

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -344,7 +345,14 @@ api.getGroups = {
     if (req.query.member) {
-      filters._id = { $in: user.guilds };
+      // Includes 'guilds' nor 'publicGuilds' type.
+      if (req.query.type === 'publicGuilds' || req.query.type === 'guilds') {
```
this could be simplified by doing something like this outside of the if block

```
if (!filters._id) {
    filters._id = {};
}

filters._id['$ne'] = TAVERN_ID;
```

so that multiple cases depending on whether we're fetching guilds or publicGuilds aren't necessary

AlexanderColen replied:
Ah I was looking for documentation on how to do that one properly with Mongoose, since I hadn't touched it before. Always nice to learn something new! 👍 

What exactly did you mean with:
> so that multiple cases depending on whether we're fetching guilds or publicGuilds aren't necessary

The documentation for type parameter says:
> Possible values are party, guilds, privateGuilds, publicGuilds, tavern.

Did you mean with this that the tavern should never be fetched? Because I was under the impression that it could be fetched with the `tavern` type parameter. Unless I am misunderstanding the code, it will never be able to fetch the Tavern if `filters._id['$ne'] = TAVERN_ID;` would always be used to filter it out, even with the `tavern` type parameter.

paglias replied:
ah good point... I think the problem might go away when the check is moved inside Group.getGroups since you'll have to filter for TAVERN_ID in the two switch cases about `guilds` and `publicGuilds` so the tavern one should not be touched at all

paglias reviewed (changes requested):
left some comments but the direction is the right one!

AlexanderColen committed:
Resolved pull request #12438 issues.

Moved Tavern exclusion to Group model, removed Group controller back to the original and resolved test failures.

AlexanderColen committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitica into develop

AlexanderColen commented:
Ended up resolving the issues in the way that we discussed them. 😃 

Thanks for the clear explanation of what exactly the issues were so that it was an easy fix on my part. 👍

paglias commented:
@AlexanderColen I'm going to be able to look at this in about 1 week

paglias commented:
thanks @AlexanderColen ! I've added the blacksmith title to your profile and noted this PR (it'll go live by the end of the week) towards your 3rd tier
